src/main/java/hudson/remoting/AgentFailuresAnalyzer.java

```diff
@@ -0,0 +1,220 @@
+        long[] deadLocks;
+        try {
+            deadLocks = mbean.findDeadlockedThreads();
+        } catch (UnsupportedOperationException x) {
+            x.printStackTrace(writer);
+            deadLocks = null;
+        }
+        if (deadLocks != null && deadLocks.length != 0) {
+            writer.println(" Deadlock Found ");
+            ThreadInfo[] deadLockThreads = mbean.getThreadInfo(deadLocks);
+            for (ThreadInfo threadInfo : deadLockThreads) {
+                StackTraceElement[] elements = threadInfo.getStackTrace();
+                for (StackTraceElement element : elements) {
+                    writer.println(element.toString());
+                }
+            }
+        }
+        writer.println();
+        writer.flush();
+        writer.close();
+    }
+
+    /**
+     * Prints the {@link ThreadInfo} (because {@link ThreadInfo#toString()} caps out the stack trace at 8 frames)
+     *
+     * @param writer the writer to print to.
+     * @param t      the thread to print
+     * @param mbean  the {@link ThreadMXBean} to use.
+     */
+    @edu.umd.cs.findbugs.annotations.SuppressWarnings(
+            value = {"VA_FORMAT_STRING_USES_NEWLINE"},
+            justification = "We don't want platform specific"
+    )
+    public static void printThreadInfo(PrintWriter writer, ThreadInfo t, ThreadMXBean mbean) {
+        writer.printf("\"%s\" id=%d (0x%x) state=%s",
+                t.getThreadName(),
+                t.getThreadId(),
+                t.getThreadId(),
+                t.getThreadState());
+        final LockInfo lock = t.getLockInfo();
+        if (lock != null && t.getThreadState() != Thread.State.BLOCKED) {
+            writer.printf("\n    - waiting on <0x%08x> (a %s)",
+                    lock.getIdentityHashCode(),
+                    lock.getClassName());
+            writer.printf("\n    - locked <0x%08x> (a %s)",
+                    lock.getIdentityHashCode(),
+                    lock.getClassName());
+        } else if (lock != null && t.getThreadState() == Thread.State.BLOCKED) {
+            writer.printf("\n    - waiting to lock <0x%08x> (a %s)",
+                    lock.getIdentityHashCode(),
+                    lock.getClassName());
+        }
+
+        if (t.isSuspended()) {
+            writer.print(" (suspended)");
+        }
+
+        if (t.isInNative()) {
+            writer.print(" (running in native)");
+        }
+
+        writer.println();
+        if (t.getLockOwnerName() != null) {
+            writer.printf("      owned by \"%s\" id=%d (0x%x)\n",
+                    t.getLockOwnerName(),
+                    t.getLockOwnerId(),
+                    t.getLockOwnerId());
+        }
+
+        final StackTraceElement[] elements = t.getStackTrace();
+        final MonitorInfo[] monitors = t.getLockedMonitors();
+
+        for (int i = 0; i < elements.length; i++) {
+            final StackTraceElement element = elements[i];
```

Comment:
NIT(?): why not java5 style? Probably leaner.
```
for(StackTraceElement element : elements) {
``` 
instead of those two lines and the `i` index "management".